File: java/src/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez;

    Tablero(int indice) {
        if (indice >= 1)
            numCasillaCarcel = indice;
        else
            numCasillaCarcel = 1;

        casillas = new ArrayList<Casilla>();
        Casilla s = new Casilla("Salida");
        casillas.add(s);

        porSalida = 0;

        tieneJuez = false;

    }

    private boolean correcto() {
        return numCasillaCarcel < casillas.size();
    }

    private boolean correcto(int numCasilla) {
        if (numCasilla >= 0 && numCasilla < casillas.size())
            return correcto();
        else
            return false;
    }

    int getCarcel() {
        return numCasillaCarcel;
    }

    int getPorSalida() {
        if (porSalida > 0) {
            porSalida--;
            return (porSalida + 1);
        } else
            return porSalida;
    }

    void añadeCasilla(Casilla casilla) {
        if (casillas.size() == numCasillaCarcel)
            casillas.add(new Casilla("Cárcel"));
        casillas.add(casilla);
        if (casillas.size() == numCasillaCarcel)
            casillas.add(new Casilla("Cárcel"));
    }

    void añadeJuez() {
        if (!tieneJuez) {
            casillas.add(new Casilla(numCasillaCarcel, "Juez"));
            tieneJuez = true;
        }
    }

    Casilla getCasilla(int numCasilla) {
        if (correcto(numCasilla))
            return casillas.get(numCasilla);
        else
            return null;
    }

    int nuevaPosicion(int actual, int tirada) {
        if (!correcto())
            return -1;
        else {
            int posicion = (actual + tirada) % casillas.size();
            if (posicion != (actual + tirada))
                porSalida++;
            return posicion;
        }
    }

    int calcularTirada(int origen, int destino) {
        int resultado = destino - origen;
        if (resultado < 0)
            resultado += casillas.size();
        return resultado;
    }

}

File: ruby/lib/tablero.rb
require_relative 'casilla'
require_relative 'casilla_calle'
require_relative 'casilla_impuesto'
require_relative 'casilla_juez'
require_relative 'casilla_sorpresa'


module Civitas
    class Tablero
        def initialize(indexCarcel)
        @numCasillaCarcel = indexCarcel >= 1 ? indexCarcel : 1
        @casillas = []
        @porSalida = 0
        @tieneJuez = false
        
        @casillas.push(Casilla.new('Salida'))

        end
        
        def correcto(numCasilla = nil)
            if (numCasilla == nil)
                return @casillas.length > @numCasillaCarcel
            elsif (@casillas.length > @numCasillaCarcel && numCasilla >= 0 && numCasilla < @casillas.length)
                return true
            end
        end

        def getCarcel()
            return @numCasillaCarcel
        end

        def getPorSalida()
            ret = @porSalida
            if (@porSalida > 0) 
                @porSalida -= 1
            end
            return ret

        end

        def añadeCasilla(casilla)
            if (@casillas.length == @numCasillaCarcel) 
                @casillas.push(Casilla.new("Cárcel"))
            end
            @casillas.push(casilla)
            if (@casillas.length == @numCasillaCarcel) 
                @casillas.push(Casilla.new("Cárcel"))
            end
        end

        def añadeJuez()
            if (!@tieneJuez)
                @casillas.push(CasillaJuez.new(@numCasillaCarcel))
                @tieneJuez = true
            end
        end

        def getCasilla(numCasilla)
            if (correcto(numCasilla))
                return @casillas[numCasilla]
            else
                return nil
            end
        end

        def nuevaPosicion(actual, tirada)
            puts "ACTUAL: #{actual}"
            puts "TIRADA: #{tirada}"
            if (correcto) 
                posicion=(actual + tirada) % @casillas.length
                if (posicion != (actual+tirada))
                    @porSalida+=1
                end
                return posicion
            end
            
            return -1
        end

        def calcularTirada(origen, destino)
            tirada = destino - origen

            if (tirada < 0)
                tirada += @casillas.length
            end

            return tirada
        end

        def toString
            string="Casilla cárcel: "+@numCasillaCarcel.to_s
            #string+=", Casillas: "+@casillas.to_s
            string+=", Juez: "+@tieneJuez.to_s
            string+=", por Salida: "+@porSalida.to_s
            #La salida de lo comentado no es ortodoxa. Activar si realmente necesario
        end



    end
end